Rework outline / table of contents

mejo- opened the pull request:
### 📝 Summary

See https://github.com/nextcloud/text/issues/7741#issuecomment-3607495936

* Resolves: #3586 
* Resolves: #5379
* Resolves: #6876
* Resolves: #7228 
* Resolves: #7522 
* Resolves: #7954 
* Resolves: nextcloud/collectives#1624 
* Resolves: nextcloud/collectives#1715 
* Resolves: nextcloud/collectives#1777

## Screenshots desktop

Outline | Outline zoomed | Table of contents | Table of contents zoomed
---|--- | --- | ---
<img width="1794" height="1174" alt="image" src="https://github.com/user-attachments/assets/22763696-8d65-4428-a85d-6cb334bc4aec" /> | <img width="360" height="1059" alt="image" src="https://github.com/user-attachments/assets/ab4f3f2a-48ba-4d34-ae90-7ec84c62435a" /> | <img width="1794" height="1174" alt="image" src="https://github.com/user-attachments/assets/5a665589-8991-4f33-803d-49cfcd7e4530" /> | <img width="444" height="1052" alt="image" src="https://github.com/user-attachments/assets/2d95bb1b-e355-43f3-ba79-61a29081f48a" />

## Screenshots mobile

Menu | Table of contents
--- | ---
<img width="617" height="1371" alt="image" src="https://github.com/user-attachments/assets/384cedd7-a582-42d9-9f49-425179c5b70a" /> | <img width="617" height="1371" alt="image" src="https://github.com/user-attachments/assets/d40dd1a7-9db9-4a60-88f2-7d77975bcbd7" />

## Screencast desktop

https://github.com/user-attachments/assets/c7c41ca9-c5fa-4296-9054-f027894d26c8

## Screencast mobile

https://github.com/user-attachments/assets/1cce768f-430d-4c12-aa4f-b736a7d894fb

### 🚧 TODO

- [x] Outline not displayed in view mode in Collectives
- [x] RichWorkspace: Hide outline action in headings menu
- [x] Update headings only when a heading changed
- [x] Update headings also in MarkdownContentEditor
- [x] Somehow `useTableOfContents` flag from `useEditorFlags` is always false in `useEditorWidth` when loading Collectives
- [x] Fix and extend tests
- [x] Test manually in a lot of scenarios:
  - [x] text standalone, collectives, (embedded?)
  - [x] collectives landing page with widgets above
  - [x] wide screen desktop, narrow screen desktop, mobile
  - [x] full width view and not 
  - [x] sidebar open and not 
  - [x] public share
  - [x] print view
  - [x] long headings
  - [x] long heading list (vertical scroll)
  - [x] conflict handling view
  - [x] read-only files

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits
- [ ] [Tests](https://github.com/nextcloud/text#-testing-the-app) (unit, integration and/or end-to-end) passing and the changes are covered with tests
- [x] Documentation ([README](https://github.com/nextcloud/text/blob/main/README.md) or [documentation](https://github.com/nextcloud/documentation/blob/master/admin_manual/configuration_server/text_configuration.rst#L2)) has been updated or is not required

marcoambrosini reviewed (commented):
Really nice @mejo- :) 

One point from my side: 

* I would trigger the outline on **hover**, not **click**
* Once hovered, I would provide a [**pin button**](https://fonts.google.com/icons?selected=Material+Symbols+Outlined:keep:FILL@0;wght@400;GRAD@0;opsz@24&icon.query=pin&icon.size=24&icon.color=%231f1f1f) in the same location where the `x` button is right now. Once the pin button is pressed, we can then display the `x` button to unpin

jancborchardt reviewed (commented):
Agree with @marcoambrosini about triggering it on hover. And when leaving the area, it should close again.

Instead of automatically staying open and needing to click a "close" button, we can have a "pin" icon that can be toggled, as proposed before. :)

mejo- commented:
> I would trigger the outline on hover, not click

Thanks for the feedback. I was unsure about that one, but now that I implemented it, I agree it's much better. I updated the desktop screencast, please have look :relaxed:

marcoambrosini reviewed (approved):
🚀

max-nextcloud commented on src/components/Editor/TableOfContents.vue:
```diff
@@ -121,67 +169,69 @@ export default {
-		&--6 {
-			--padding-left: 5rem;
+		&.level6 {
+			--padding-inline-start: calc(5 * var(--level-padding));
```
I think you could use an scss for loop here.
https://sass-lang.com/documentation/at-rules/control/for/

mejo- commented on src/components/Editor.vue:
```diff
@@ -522,6 +527,9 @@ export default defineComponent({
 			this.updateUser(session)
+			this.updateHeadings(
+				[...headingAnchorPluginKey.getState(this.editor.state)?.headings] ?? []
+			)
```
Sorry, this was refactored in a later commit. Now `editor` is passed as argument to `useEditorHeadings()`.

mejo- commented on src/components/Editor/ContentContainer.vue:
```diff
@@ -73,40 +66,13 @@ export default {
 .editor__content {
-	max-width: min(var(--text-editor-max-width), calc(100vw - 16px));
+	max-width: min(var(--text-editor-max-width), 100vw);
+	width: min(var(--text-editor-max-width), 100vw);
```
Same here, I removed `width` again in a subsequent commit.

max-nextcloud commented on src/composables/useEditorWidth.ts:
```diff
@@ -43,11 +44,13 @@ export const editorWidthKey = Symbol('text:editor:width') as InjectionKey<
-const defaultEditorWidthMobile = 'min(80ch, 100%)'
-const fullEditorWidthDesktop = 'calc(100% - 2 * 40px)'
+const editorWidthDesktop = 'min(80ch, 100%)'
+const editorWidthDesktopEnhanced = 'min(80ch, (100% - 2 * 40px))'
```
I wonder if we could express this in css - as in using some sibling selector.

mejo- commented on src/editor.js:
```diff
@@ -190,7 +188,8 @@ window.OCA.Text.createEditor = async function ({
 	onLoaded = () => {},
 	onUpdate = ({ markdown }) => {},
-	onOutlineToggle = (visible) => {},
+	onTocToggle = (visible) => {},
```
Well, I changed it to `onTocToggle` everywhere now, given that we have `onOutlineToggle` (without the `d`) already.
